text/0071-trailing-commas.md

```diff
@@ -0,0 +1,170 @@
+# Trailing Commas
+
+## Related issues and PRs
+
+- Reference Issues: https://github.com/cedar-policy/cedar/issues/1024
+- Implementation PR(s):
+
+## Timeline
+
+- Started: 2024-06-26 
+
+## Summary
+
+The Cedar grammar (both policies and schemas) should accept trailing
+commas whenever possible.
+
+## Basic example
+
+Currently, this is a parse error:
+```
+permit(principal, action, resource) when {
+    { foo : 3, bar : 2, } == context  
+};
+```
+This RFC would make it not a parse error. 
+It would also change this is several places beyond records.
+
+## Motivation
+
+In general, allowing trailing commas makes editing easier.
+Re-arranging, deleting, and copy-pasting code into/from a comma
+separated grammar element will no longer require adjusting commas.
+
+Example:
+```
+{
+    "managers": ["Dave", "Shelly", "Chris"],
+    "projects" : {
+        "Alpha" : {
+            "zones" : ["A", "B"]
+        },
+        "Beta" : {
+            "zones" : ["D", "F"]
+        }
+    }
+}
+```
+The `Beta` project cannot simply be deleted from this record, and
+likewise, new fields cannot be simply pasted into any level of the
+objects, without adjusting commas.
+
+Many languages allow for trailing commas, including:
+* Rust
+* Python
+* Java,
+* Go
+* JavaScript
+* OCaml
+
+(Frustratingly: not JSON!)
+
+In addition, it will make our schema grammar and policy grammar more
+consistent, as the schema grammar allows trailing commas in many places.
+
+## Detailed design
```

Comment:
It'd be good to list out an example for every expression where we'd support it before digging in to the grammar.

What comes to mind is:

```
permit(principal, action, resource, )'
[1, 2, ]
{a: 1, b: 2, }
ip("0.0.0.0", )
principal.tags.contains("x", )
```

Anything else?

Annotations feel like they could also take a trailing comma if we wanted to stretch the limits of this RFC. Something like `@id("permit-admin", )`

Reply:
My only argument against doing it for annotations is it implies you _could_ supply more than one arg: which you can't.